### PR TITLE
Intellij e2e

### PR DESCRIPTION
**Pull request by infiniteregrets**

### CODE REVIEW

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+                        }
+                    }
+                    statusBar {
+                        // need to make sure poetry is not doing anything
+                        try {
+                            waitFor {
+                                !poetryProgress.isShowing
+                            }
+                        } catch (e: Exception) {
+                            // ignore
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:137]: Why are you ignoring this?

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                        waitFor {
+                            isFileOpened("app.py")
+                        }
+                    }
+                }
+                step("Set up Poetry Environment") {
+                    // blue hover appears on top of the text window asking
+                    // to set up poetry environment
+                    fileIntention {
+                        val setUpPoetry = setUpPoetry
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:124]: Can you explain this line?

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                                !poetryProgress.isShowing
+                            }
+                        } catch (e: Exception) {
+                            // ignore
+                        }
+                    }
+                }
+
+                step("Enable mirrord and create config file") {
+                    waitFor(ofSeconds(30)) {
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:143]: Why are you batching them? Makes more sense to test the config first, see if it broke, and if it didn't move forward

File: src/test/kotlin/com/metalbear/mirrord/utils/DialogFixture.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/utils/DialogFixture.kt:2]: Why is this file necessary?

File: src/test/kotlin/com/metalbear/mirrord/utils/IdeaFrame.kt [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        )
+
+    val debuggerConnected
+        get() = find<ContainerFixture>(byXpath("//div[@class='XDebuggerTree']"))
+
+    val xDebuggerFramesList
+        get() = find<ContainerFixture>(byXpath("//div[@class='XDebuggerFramesList']"))
+
+    @JvmOverloads
+    fun dumbAware(timeout: Duration = Duration.ofMinutes(5), function: () -> Unit) {
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/utils/IdeaFrame.kt:60]: Might be a good place to document what smart and dumb mode is

File: src/test/kotlin/com/metalbear/mirrord/utils/StepsLogger.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.intellij.remoterobot.stepsProcessing.StepLogger
+import com.intellij.remoterobot.stepsProcessing.StepWorker
+
+object StepsLogger {
+    private var initializaed = false
+
+    @JvmStatic
+    fun init() {
+        if (initializaed.not()) {
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/utils/StepsLogger.kt:11]:
```suggestion
        if (initialized.not()) {
```

File: src/test/kotlin/com/metalbear/mirrord/utils/WelcomeFixture.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/utils/WelcomeFixture.kt:2]: Why is this file necessary?

**Review by eyalb181 — CHANGES_REQUESTED** [review]
Well done! Made some comments

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 1 above)
Comment (infiniteregrets) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:137]: so the issue is we dont want to do anything if intellij is doing anything with poetry setup, same as indexing so i try to check if its happening then just ignore if it times out that means nothing is in the status bar related to poetry

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 2 above)
Comment (infiniteregrets) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:124]: there is a blue bar on the text window when a python file is opened in a poetry project, we click on a button in it to setup poetry quickly

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 3 above)
Comment (infiniteregrets) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:143]: not sure what you mean? do you mean separate the button clicks?

File: src/test/kotlin/com/metalbear/mirrord/utils/DialogFixture.kt (hunk 4 above)
Comment (infiniteregrets) [reply src/test/kotlin/com/metalbear/mirrord/utils/DialogFixture.kt:2]: represents a dialog like mirrord dialog asking to choose pod

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 2 above)
Comment (eyalb181) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:125]: I meant more like, why do you need `val setUpPoetry`? Can't you just do `setUpPoetry.click()` directly?

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 1 above)
Comment (eyalb181) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt]: But your waitFor clause is waiting for poetryProgress *not* to show, so timeout means that poetryProgress is still showing, right?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 2 above)
Comment (infiniteregrets) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:125]: this is because it is a getter function: 
```
val setUpPoetry
        get() = find<ContainerFixture>(
            byXpath("//div[@accessiblename.key='sdk.set.up.poetry.environment']"),
            Duration.ofSeconds(20)
        )
```
i need that value again in the `waitFor`

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 1 above)
Comment (infiniteregrets) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt]: yes correct, i think i should remove the try catch

Comment (infiniteregrets) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt]: thanks i have removed it

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt (hunk 3 above)
Comment (eyalb181) [reply src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:140]: I mean separate steps - so you'd have a step for creating the config file, and we'd know that if that step failed, there's an issue with the config file creation.

Here's how I'd split it into steps:
1. create config file (this is mirrord functionality so should be tested)
2. open app.py, set breakpoint (this is just intellij functionality so nothing to test and can be batched)
3. enable mirrord, start debugging (enable mirrord shouldn't have its own step since it doesn't really have any functionality to test until you start debugging)

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt [hunk 8]
```diff
@@ -90,23 +94,7 @@ internal class MirrordPluginTest {
...
-                }
                 with(projectViewTree) {
-                    if (clickToOpenProject) {
-                        waitFor(ofSeconds(30)) {
-                            hasText("test-workspace")
-                        }
-                        findText("test-workspace").doubleClick()
-                    }
                     waitFor(ofSeconds(30)) {
                         hasText("app.py")
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:99]: Why is this necessary to verify at this stage?

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt [hunk 9]
```diff
@@ -130,28 +118,8 @@ internal class MirrordPluginTest {
...
-                    }
-                    editorTabs {
-                        waitFor {
-                            isFileOpened("mirrord.json")
-                        }
-                        findText("app.py").click()
-                        waitFor {
-                            isFileOpened("app.py")
+                        waitFor(ofSeconds(120)) {
+                            isProgressBarEmpty()
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:122]: Why is this necessary to verify at this stage?

File: src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt [hunk 10]
```diff
@@ -168,8 +136,32 @@ internal class MirrordPluginTest {
...
+                }
+                dumbAware {
+                    createMirrordConfig.click()
+                }
+                editorTabs {
+                    waitFor {
+                        isFileOpened("mirrord.json")
+                    }
+                    // switch back to app.py
+                    findText("app.py").click()
```
Comment (eyalb181) [inline src/test/kotlin/com/metalbear/mirrord/MirrordPluginTest.kt:151]: This shouldn't be done here, but rather just when you need it - before setting the breakpoint

**Review by eyalb181 — CHANGES_REQUESTED** [review]

**Review by eyalb181 — APPROVED** [review]
